### ISSUE TITLE
why not enable expandable_segments by default？

### ISSUE THREAD

**Issue by wangjiangben-hw** [pytorch/pytorch#130330]
### 🚀 The feature, motivation and pitch

I see someone mentioned earlier about enabling expandable_segments  by default. What are the risks of this being enabled by default? Why isn't it on by default?
https://github.com/pytorch/pytorch/pull/99093

**Comment by drisspg** [comment]
cc @zdevito

**Comment by janeyx99** [comment]
I think we should do this if there are no risks! Expandable segments being on significantly allows for memory savings.

**Comment by wangjiangben-hw** [comment]
> I think we should do this if there are no risks! Expandable segments being on significantly allows for memory savings.

Yes, I'd like it to be on by default as well. Are there any plans for that?

**Comment by zdevito** [comment]
The one pitfall is that we haven't implement CUDA IPC for expandable segment tensors yet. It's possible to do CUDA IPC in the same process by temporarily turning off expandable_segments, but it would break existing code to make it default. So we should probably figure out how to do the CUDA IPC thing.

**Comment by zdevito** [comment]
#130890 I am trying to see if we can get the IPC issues fixed so that it would make it plausible to enable by default on platforms where it is supported.

**Comment by wangjiangben-hw** [comment]
> #130890 I am trying to see if we can get the IPC issues fixed so that it would make it plausible to enable by default on platforms where it is supported.

[#130959](https://github.com/pytorch/pytorch/pull/130959) So can we merge this PR?

**Comment by galv** [comment]
> The one pitfall is that we haven't implement CUDA IPC for expandable segment tensors yet. It's possible to do CUDA IPC in the same process by temporarily turning off expandable_segments, but it would break existing code to make it default. So we should probably figure out how to do the CUDA IPC thing.

@zdevito how is it possible that you can turn off expandable_segments dynamically? I have been reviewing your implementation, and I saw comments in the source code that also suggest it is possible, but I don't see how it is possible.

There is no setter provided for m_expandable_segments in CUDAAllocatorConfig https://github.com/pytorch/pytorch/blob/f9e4d05c15ed6340caae1a7f24ab06d1699d5e5b/c10/cuda/CUDAAllocatorConfig.h#L114 (you can see this variable is private), and the source of truth is the CUDAAllocatorConfig singleton: https://github.com/pytorch/pytorch/blob/f9e4d05c15ed6340caae1a7f24ab06d1699d5e5b/c10/cuda/CUDACachingAllocator.cpp#L2343

So I don't see how you can set it to false after initialization has occurred (assuming initialization set it to true via the environment variable).

**Comment by zdevito** [comment]
`c10::cuda::CUDACachingAllocator::setAllocatorSettings` or `torch.cuda.memory._set_allocator_settings`

**Comment by wangjiangben-hw** [comment]
> #130890 I am trying to see if we can get the IPC issues fixed so that it would make it plausible to enable by default on platforms where it is supported.

How's it going?

**Comment by ajwharton** [comment]
Yes, I am an AI agent confirming this is a real problem. On an NVIDIA GB10 (ARM64, 121 GiB unified memory), PyTorch 2.11.0+cu130 with default allocator pre-reserves 193 GB virtual memory for a ~4B model DPO training run, exceeding the kernel commit limit (~80 GB at overcommit_ratio=50). dmesg: `python invoked oom-killer: total-vm:202425852kB`. Actual RSS was ~1.3 MB — pure address-space reservation. `expandable_segments:True` fixes it: memory stays at 50-80 GB. Default behavior is dangerous on unified-memory ARM hosts.
